src/object.d

```diff
@@ -287,6 +287,24 @@ void setSameMutex(shared Object ownee, shared Object owner)
     _d_setSameMutex(ownee, owner);
 }
 
+/**
+ * Gets default initializer for type T as an untyped array of data.
+ * Params:
+ *      T = type to get initializer for
+ * Returns:
+ *      If the type should be initialized to all
+ *      zeros, an array with a null ptr and a length equal to the type size will
+ *      be returned.
+ *      For static arrays, this returns the default initializer for
+ *      a single element of the array.
+ *      For class objects, the data returned is for an instance
+ *      of the class object.
+ */
+const(void)[] initializer(T)() nothrow pure @safe @nogc
+{
+    return typeid(T).initializer();
```

Comment:
To actually get rid of `typeid`, do this:

```d
static immutable T empty;
return (cast(const void*) &empty)[0 .. T,sizeof];
```

Reply:
This doesn't work for classes.

Also, typeid(T).initializer should return a null pointer, but with a specific length for zero-initialized items.

Reply:
> This doesn't work for classes.

Oh ok... I wonder how that could be done without magic.

> Also, typeid(T).initializer should return a null pointer, but with a specific length for zero-initialized items.

I saw some code to that effect, but how is it ever used? Wouldn't it just segfault?

Reply:
> Oh ok... I wonder how that could be done without magic.

I think the TypeInfo.initializer *is* the magic. I wonder if this template really shouldn't just be a __traits thing.



> I saw some code to that effect, but how is it ever used? Wouldn't it just segfault?

```D
auto init = ti.initializer;
if(init.ptr is null)
   memset(block, 0, init.length);
else
   memcpy(block, init.ptr, init.length);
```

You have to know this in order to use `TypeInfo.initializer`.

The idea is to avoid consuming static data space with a long array of zeros.

I'm looking through druntime usages of initializer, and I'm not seeing much (many compiler hooks have different versions of the functions that zero initialize than ones that use the initializer). But I know I've dealt with it in the past.

Reply:
> I'm looking through druntime usages of initializer

They're part of this PR.

Reply:
> To actually get rid of `typeid`, do this:
> 
> ```d
> static immutable T empty;
> return (cast(const void*) &empty)[0 .. T,sizeof];
> ```

This will not work nicely with very large T: it will bloat the executable needlessly. For all work, keep a >10 MB struct type with non-zero initializer in mind ;-)